src/java/org/apache/cassandra/index/sai/utils/TermIterator.java

```diff
@@ -0,0 +1,138 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.exceptions.QueryCancelledException;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.index.sai.SSTableIndex;
+import org.apache.cassandra.index.sai.SSTableQueryContext;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.Throwables;
+
+public class TermIterator extends KeyRangeIterator
+{
+    private static final Logger logger = LoggerFactory.getLogger(TermIterator.class);
+
+    private final QueryContext context;
+
+    private final KeyRangeIterator union;
+    private final Set<SSTableIndex> referencedIndexes;
+
+    private TermIterator(KeyRangeIterator union, Set<SSTableIndex> referencedIndexes, QueryContext queryContext)
+    {
+        super(union.getMinimum(), union.getMaximum(), union.getCount());
+
+        this.union = union;
+        this.referencedIndexes = referencedIndexes;
+        this.context = queryContext;
+    }
+
+    @SuppressWarnings({"resource", "RedundantSuppression"})
+    public static TermIterator build(final Expression e,
+                                     Set<SSTableIndex> perSSTableIndexes,
+                                     AbstractBounds<PartitionPosition> keyRange,
+                                     QueryContext queryContext)
+    {
+        final List<KeyRangeIterator> tokens = new ArrayList<>(1 + perSSTableIndexes.size());
+
+        KeyRangeIterator memtableIterator = e.context.searchMemtableIndexes(e, keyRange);
+        if (memtableIterator != null)
+            tokens.add(memtableIterator);
+
+        for (final SSTableIndex index : perSSTableIndexes)
+        {
+            try
+            {
+                queryContext.checkpoint();
+                queryContext.incSstablesHit();
+                assert !index.isReleased();
+
+                SSTableQueryContext context = queryContext.getSSTableQueryContext(index.getSSTable());
+                List<KeyRangeIterator> keyIterators = index.search(e, keyRange, context);
```

Comment:
```suggestion
                List<KeyRangeIterator> segmentIterators = index.search(e, keyRange, context);
```
Also, `search()` JavaDoc could probably explain it's a list *because* of the segments